openSearchLane.go

```diff
@@ -0,0 +1,425 @@
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/opensearch-project/opensearch-go/v3"
+	"github.com/opensearch-project/opensearch-go/v3/opensearchapi"
+)
+
+const (
+	DefaultLogThreshold    = 5
+	DefaultMaxBufferSize   = 10
+	DefaultBackoffInterval = 10 * time.Second
+	DefaultBackOffLimit    = 10 * time.Minute
+)
+
+type OpenSearchLogMessage struct {
+	AppName      string            `json:"appName"`
+	ParentLaneId string            `json:"parentLaneId,omitempty"`
+	JourneyID    string            `json:"journeyId,omitempty"`
+	LaneID       string            `json:"laneId,omitempty"`
+	LogMessage   string            `json:"logMessage,omitempty"`
+	Metadata     map[string]string `json:"metadata,omitempty"`
+}
+
+type EmergencyFn func(logBuffer []*OpenSearchLogMessage)
+
+type OpenSearchLaneConfig struct {
+	OpenSearchUrl     string        `json:"openSearchUrl"`
+	OpenSearchPort    string        `json:"openSearchPort"`
+	OpenSearchUser    string        `json:"openSearchUser"`
+	OpenSearchPass    string        `json:"openSearchPass"`
+	OpenSearchIndex   string        `json:"openSearchIndex"`
+	OpenSearchAppName string        `json:"openSearchAppName"`
+	LogThreshold      int           `json:"logThreshold,omitempty"`
+	MaxBufferSize     int           `json:"maxBufferSize,omitempty"`
+	BackoffInterval   time.Duration `json:"backoffInterval,omitempty"`
+	BackOffLimit      time.Duration `json:"backoffLimit,omitempty"`
+}
+
+type openSearchLane struct {
+	logLane
+	openSearchConnection *openSearchConnection
+	metadata             map[string]string
+}
+
+type openSearchConnection struct {
+	client         *opensearchapi.Client
+	mu             sync.Mutex
+	logBuffer      []*OpenSearchLogMessage
+	stopCh         chan *sync.WaitGroup
+	wakeCh         chan struct{}
+	refCount       int
+	config         OpenSearchLaneConfig
+	emergencyFn    EmergencyFn
+	ctx            context.Context
+	cancelFn       context.CancelFunc
+	messagesQueued int
+	messagesSent   int
+}
+
+type OpenSearchLane interface {
+	Lane
+	LaneMetadata
+	Connect(config OpenSearchLaneConfig) (err error)
+	SetEmergencyHandler(emergencyFn EmergencyFn)
+	Stats()
+}
+
+func NewOpenSearchLane(ctx context.Context) (l OpenSearchLane) {
+	ll := deriveLogLane(nil, ctx, []Lane{}, "")
+
+	ctx, cancelFn := context.WithCancel(context.Background())
+
+	osl := openSearchLane{
+		openSearchConnection: &openSearchConnection{
+			logBuffer: make([]*OpenSearchLogMessage, 0),
+			stopCh:    make(chan *sync.WaitGroup, 1),
+			wakeCh:    make(chan struct{}, 1),
+			ctx:       ctx,
+			cancelFn:  cancelFn,
+		},
+	}
+	osl.openSearchConnection.attach()
+
+	ll.clone(&osl.logLane)
+	osl.logLane.writer = log.New(&osl, "", 0)
+	osl.wlog.SetFlags(log.Flags() &^ (log.Ldate | log.Ltime))
+
+	l = &osl
+
+	return
+
+}
+
+func (osl *openSearchLane) Connect(config OpenSearchLaneConfig) (err error) {
+	if osl.openSearchConnection.client != nil {
```

Comment:
Instead of erroring, it should tell the opensearchconneciton to shut its client and reopen it with the different config. This is needed in order to support hot config loading.

Reply:
Maybe use the wakeCh for this purpose.

Reply:
done, opensearch connection doesn't have an active client connection so it will just be overwritten with the new one